Implement signTransaction

paweljakubas opened the pull request:
### Overview

- [x] fix output when constructing tx and show it works in integration tests (adp-909 artefact)
- [x] extend Transaction layer by adding `mkSignedTransaction`
- [x] add impl of signing in Shelley.Transaction
- [x] handle deserializing of value of SerializedTx
- [x] add signTx as link and extend integration test 
- [x] handle extracting inputs from TxBody
- [x] handle extracting withdrawals from TxBody
- [x] do Shelley.Transaction refactoring 
- [x] align Wallet.signTransaction accordingly
- [x] align Api.Server.signTransaction accordingly

### Comments

Based on the branch of PR #2877. Merge that first.

### Issue Number

ADP-919

paweljakubas committed:
Implement Wallet.signTransaction

Co-authored-by: Pawel Jakubas <pawel.jakubas@iohk.io>
Co-authored-by: Rodney Lorrimar <rodney.lorrimar@iohk.io>

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -202,45 +206,55 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
             [ expectSuccess
             , expectResponseCode HTTP.status202
-            , expectField (#coinSelection . #inputs) (`shouldSatisfy` (not . null))
-            , expectField (#coinSelection . #outputs) (`shouldSatisfy` (not . null))
```
Why removing this assertion, especially that the test was failing before due to missing outputs? (pls see `pendingWith` message)

paweljakubas replied:
just check it later in the same test, but deeper:
```
        let coinSelOutputs = getFromResponse (#coinSelection . #outputs) rConstrTx
        coinSelOutputs `shouldBe`
            [ApiCoinSelectionOutput destAddr (Quantity amt) (ApiT W.empty)]
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3299,6 +3301,40 @@ instance IsServerError ErrSignPayment where
+                [ "Whoops, it seems like I just experienced a hard-fork in the "
+                , "middle of other tasks. This is a pretty rare situation but "
+                , "as a result, I must throw-away what I was doing. Please "
+                , "retry whatever you were doing in a short delay."
```
```suggestion
                , "as a result, I must throw away what I was doing. Please "
                , "retry your request."
```

Hi @paweljakubas: I'm just wondering, how long should the user wait here? Is it a case of a few seconds, or a few minutes?

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -392,11 +416,91 @@ instance Buildable a => Buildable (WithDirection a) where
+withinEra (AnyCardanoEra ShelleyEra) (AnyCardanoEra AllegraEra) = False
+withinEra (AnyCardanoEra ShelleyEra) (AnyCardanoEra _) = True
+withinEra (AnyCardanoEra ByronEra) (AnyCardanoEra ByronEra) = True
+withinEra (AnyCardanoEra ByronEra) _ = False
```
Hi @paweljakubas. Just wondering, could this be expressed in terms of ordering for eras, with `<`?

Alternatively, this seems a bit shorter:

```hs
withinEra :: AnyCardanoEra -> AnyCardanoEra -> Bool
withinEra (AnyCardanoEra e1) (AnyCardanoEra e2) = case (e1, e2) of
    (AlonzoEra , _         ) -> True
    (MaryEra   , AlonzoEra ) -> False
    (MaryEra   , _         ) -> True
    (AllegraEra, AlonzoEra ) -> False
    (AllegraEra, MaryEra   ) -> False
    (AllegraEra, _         ) -> True
    (ShelleyEra, AlonzoEra ) -> False
    (ShelleyEra, MaryEra   ) -> False
    (ShelleyEra, AllegraEra) -> False
    (ShelleyEra, _         ) -> True
    (ByronEra  , ByronEra  ) -> True
    (ByronEra  , _         ) -> False
```

(just thinking out loud: maybe it's also nice to avoid `_` here, to be more explicit about which cases match, and catch compile errors in case this sum type gets changed)

rvl replied:
There is no `Ord` on `AnyCardanoEra`, so that's what we're defining here.
The wildcard are because no era is higher that AlonzoEra, as an example.
Anyway I think it can be made shorter just by assigning a natural to each era.

HeinrichApfelmus replied:
Assigning naturals would be morally equivalent to treating `AnyCardanoera` as an `Enum`, I suppose?

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -465,18 +465,19 @@ cardanoTxFromBytes maxEra bs = asum $ map snd $ filter (withinEra maxEra . fst)
+    (ShelleyEra, AllegraEra) -> False
+    (ShelleyEra, _         ) -> True
+    (ByronEra  , ByronEra  ) -> True
+    (ByronEra  , _         ) -> False
```
Hey @paweljakubas. I wonder if this is possible to express with `>=`.

Is there an `Ord` instance for the type of value within `AnyCardanoEra`?

If so, it seems like the case expression could be replaced with something like `x >= y`.

But perhaps I've misunderstood something here. 😄

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -392,11 +416,92 @@ instance Buildable a => Buildable (WithDirection a) where
+    , (anyCardanoEra ShelleyEra, InAnyCardanoEra ShelleyEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsShelleyEra) bs)
+    , (anyCardanoEra ByronEra, InAnyCardanoEra ByronEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsByronEra) bs)
+    ]
+  where
```
Suggestion: extract out the shared logic into a separate function:

```suggestion
cardanoTxFromBytes maxEra bs = asum $ map snd $ filter (withinEra maxEra . fst)
    [ deserialise AlonzoEra  Cardano.AsAlonzoEra
    , deserialise MaryEra    Cardano.AsMaryEra
    , deserialise AllegraEra Cardano.AsAllegraEra
    , deserialise ShelleyEra Cardano.AsShelleyEra
    , deserialise ByronEra   Cardano.AsByronEra
    ]
  where
    deserialise
        :: forall era. Cardano.IsCardanoEra era
        => CardanoEra era
        -> Cardano.AsType era
        -> (AnyCardanoEra, Either DecoderError (InAnyCardanoEra Cardano.Tx))
    deserialise era asEra =
        ( anyCardanoEra era
        , InAnyCardanoEra era <$> deserialiseFromCBOR (Cardano.AsTx asEra) bs
        )
```

rvl replied:
Yep good idea, thanks.

rvl committed:
Rework, refactor, allow partial signing

Co-authored-by: Pawel Jakubas <pawel.jakubas@iohk.io>
Co-authored-by: Rodney Lorrimar <rodney.lorrimar@iohk.io>

rvl committed:
swagger: Add withdrawal to sign transaction request

rvl committed:
Update golden JSON files

rvl committed:
Skip a failing test

The test case was tied closely to the previous implementation of
signTransaction.

rvl committed:
Regenerate nix

rvl committed:
Add Test.Utils.Pretty.(====)

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Balance.hs:
```diff
@@ -510,15 +515,6 @@ performSelection minCoinFor costFor bundleSizeAssessor criteria
-        , extraCoinSource
-        , assetsToMint
-        , assetsToBurn
-        } = criteria
```
Hi @rvl.

I'd like to ask a favour: would it be possible to avoid changing these to use `RecordWildCards`, at least for the moment? 

There are two things here:
- it makes it harder to figure out where fields have come from, as some field names are shared by multiple records. If we use record wildcards, then go-to-definition no longer works.
- I'm currently working on this module, so it would help me greatly if we could keep any changes to a minimum!

Thanks!

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -192,20 +305,20 @@ data SelectionParams = SelectionParams
     -> NonEmpty TxOut
     -> Either ErrPrepareOutputs (NonEmpty TxOut)
-prepareOutputs constraints outputsUnprepared
+prepareOutputsForMinUTxO constraints outputsUnprepared
```
Hi @rvl. I think this name change is actually unnecessary, as this function is not just about the minimum UTxO: it also:
- checks the serialized sizes of outputs
- checks that outputs don't have token quantities in excess of the maximum (`maxBound :: Word64`).

I reckon it would be better to leave the name alone here.

rvl committed:
Simplify delegation and transaction signing, rework key store

rvl committed:
Revert PR #2874

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -202,45 +206,54 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+            , expectResponseCode HTTP.status202
+            ]
+
+        -- and send it in two steps
```
So it seems this part is missing. Can this be added at this stage already?
